client/devicemanager/instance.go

```diff
@@ -0,0 +1,518 @@
+		}
+	}
+
+	// Convert to a fingerprint plugin
+	device, ok := pluginInstance.Plugin().(device.DevicePlugin)
+	if !ok {
+		pluginInstance.Kill()
+		return nil, fmt.Errorf("plugin loaded does not implement the driver interface")
+	}
+
+	// Store the plugin and device
+	i.plugin = pluginInstance
+	i.device = device
+
+	// Store the reattach config
+	if c, ok := pluginInstance.ReattachConfig(); ok {
+		i.storeReattach(c)
+	}
+
+	return device, nil
+}
+
+// cleanup shutsdown the plugin
+func (i *instanceManager) cleanup() {
+	i.shutdownLock.Lock()
+	defer i.shutdownLock.Unlock()
+
+	if i.plugin != nil && !i.plugin.Exited() {
+		i.plugin.Kill()
+		i.storeReattach(nil)
+	}
+}
+
+// fingerprint is a long lived routine used to fingerprint the device
+func (i *instanceManager) fingerprint() {
+	defer i.wg.Done()
+
+START:
+	// Get a device plugin
+	devicePlugin, err := i.dispense()
+	if err != nil {
+		i.logger.Error("dispensing plugin failed", "error", err)
+		i.cancel()
+		return
+	}
+
+	// Start fingerprinting
+	fingerprintCh, err := devicePlugin.Fingerprint(i.ctx)
+	if err != nil {
+		i.logger.Error("fingerprinting failed", "error", err)
+		i.handleFingerprintError()
+		return
+	}
+
+	var fresp *device.FingerprintResponse
+	var ok bool
+	for {
+		select {
+		case <-i.ctx.Done():
+			return
+		case fresp, ok = <-fingerprintCh:
+		}
+
+		if !ok {
+			i.logger.Debug("exiting since fingerprinting gracefully shutdown")
+			i.handleFingerprintError()
+			return
+		}
+
+		// Guard against error by the plugin
+		if fresp == nil {
+			continue
+		}
+
+		// Handle any errors
+		if fresp.Error != nil {
+			if fresp.Error == base.ErrPluginShutdown {
+				i.logger.Error("plugin exited unexpectedly")
+				goto START
+			}
+
+			i.logger.Error("fingerprinting returned an error", "error", err)
+			i.handleFingerprintError()
+			return
+		}
+
+		if err := i.handleFingerprint(fresp); err != nil {
+			// Cancel the context so we cleanup all goroutines
+			i.logger.Error("returned devices failed fingerprinting", "error", err)
+			i.handleFingerprintError()
+		}
+	}
+}
+
+// handleFingerprintError exits the manager and shutsdown the plugin.
+func (i *instanceManager) handleFingerprintError() {
+	// Clear out the devices and trigger a node update
+	i.deviceLock.Lock()
+	defer i.deviceLock.Unlock()
+
+	// If we have fingerprinted before clear it out
+	if i.hasFingerprinted {
+		// Store the new devices
+		i.devices = nil
+
+		// Trigger that the we have new devices
+		select {
+		case i.fingerprintOutCh <- struct{}{}:
+		default:
+		}
+	}
+
+	// Cancel the context so we cleanup all goroutines
+	i.cancel()
+
+	return
+}
+
+// handleFingerprint stores the new devices and triggers the fingerprint output
+// channel. An error is returned if the passed devices don't pass validation.
+func (i *instanceManager) handleFingerprint(f *device.FingerprintResponse) error {
+	// Safety check
+	if f.Devices == nil {
+		return nil
```

Comment:
Should this ever happen? If it's a programming error we should at least log so developers aren't wondering why nothing is happening if they forget to set that.

Reply:
Put a better comment. You could imagine a fingerprint returning no devices if there are none on the node.